File: elixir/day10_part1_2023.exs

defmodule Day10 do
  @empty "."
  @start "S"
  @vertical "|"
  @horizontal "-"
  @top_left_corner "J"
  @top_right_corner "L"
  @bottom_left_corner "7"
  @bottom_right_corner "F"
  @enclosed "X"

  @tile_to_pipe %{
    @vertical => %{top: true, bottom: true},
    @horizontal => %{left: true, right: true},
    @top_left_corner => %{top: true, left: true},
    @top_right_corner => %{top: true, right: true},
    @bottom_left_corner => %{bottom: true, left: true},
    @bottom_right_corner => %{bottom: true, right: true}
  }

  def solve(input) do
    grid = build_grid(input)
    start = find_start(grid)
    path = path_finding(start, grid)
    length(path) / 2
  end

  def build_grid(input) do
    input
    |> Enum.with_index()
    |> Enum.flat_map(fn {line, y} ->
      line
      |> String.graphemes()
      |> Enum.with_index()
      |> Enum.filter(fn {char, _x} -> char != @empty end)
      |> Enum.map(fn {char, x} -> {{x, y}, char} end)
    end)
    |> Map.new()
  end

  def find_start(grid) do
    grid
    |> Enum.find(fn {{_x, _y}, char} -> char == @start end)
    |> case do
      {{x, y}, _char} -> {x, y}
    end
  end

  def get_pipe_from_tile(tile) do
    Map.get(@tile_to_pipe, tile, %{})
  end

  def get_tile_from_pipe(pipe) do
    @tile_to_pipe
    |> Enum.find(fn {_tile, associated_pipe} -> pipe == associated_pipe end)
    |> case do
      {tile, _pipe} -> tile
      nil -> @empty
    end
  end

  def get_pipe_from_neighbors({x, y}, grid) do
    possible_neighbors = %{
      top: {x, y - 1},
      right: {x + 1, y},
      bottom: {x, y + 1},
      left: {x - 1, y}
    }

    possible_neighbors
    |> Enum.reduce(%{}, fn {dir, {nx, ny}}, acc ->
      case Map.get(grid, {nx, ny}) do
        nil ->
          acc

        tile ->
          neighbor_pipe = get_pipe_from_tile(tile)

          if Map.has_key?(neighbor_pipe, opposite_dir(dir)) do
            Map.put(acc, dir, true)
          else
            acc
          end
      end
    end)
  end

  def opposite_dir(dir) do
    case dir do
      :top -> :bottom
      :right -> :left
      :bottom -> :top
      :left -> :right
    end
  end

  def path_finding(start, grid) do
    start_pipe = get_pipe_from_neighbors(start, grid)
    {previous_dir, _} = Enum.find(start_pipe, fn {_, _} -> true end)
    current = move(start, previous_dir)
    path = [start, current]

    path_finding(current, previous_dir, start, path, grid)
  end

  def path_finding(current, previous_dir, start, path, grid) do
    if current == start do
      path
    else
      current_pipe = get_pipe_from_tile(Map.get(grid, current))

      {new_dir, _} =
        current_pipe
        |> Enum.find(fn {dir, _} -> dir != opposite_dir(previous_dir) end)

      next_coord = move(current, new_dir)
      path_finding(next_coord, new_dir, start, path ++ [next_coord], grid)
    end
  end

  def move({x, y}, dir) do
    case dir do
      :top -> {x, y - 1}
      :right -> {x + 1, y}
      :bottom -> {x, y + 1}
      :left -> {x - 1, y}
    end
  end

  def read_file(filename) do
    File.read!(filename)
    |> String.trim()
    |> String.split("\n")
  end
end

input = Day10.read_file("input.txt")
IO.puts(Day10.solve(input))

File: cpp/day10_part1_2023.cpp

#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <utility>
#include <vector>

using Coord = std::pair<int, int>;
using PipeDirections = std::set<Coord>;

const char EMPTY = '.';
const char START = 'S';
const char VERTICAL = '|';
const char HORIZONTAL = '-';
const char TOP_LEFT_CORNER = 'J';
const char TOP_RIGHT_CORNER = 'L';
const char BOTTOM_LEFT_CORNER = '7';
const char BOTTOM_RIGHT_CORNER = 'F';

const Coord UNDEFINED = {0, 0};
const Coord TOP = {0, -1};
const Coord RIGHT = {1, 0};
const Coord BOTTOM = {0, 1};
const Coord LEFT = {-1, 0};

std::map<char, PipeDirections> TILE_TO_PIPE;

void initTileToPipe() {
    TILE_TO_PIPE[VERTICAL] = {TOP, BOTTOM};
    TILE_TO_PIPE[HORIZONTAL] = {LEFT, RIGHT};
    TILE_TO_PIPE[TOP_LEFT_CORNER] = {TOP, LEFT};
    TILE_TO_PIPE[TOP_RIGHT_CORNER] = {TOP, RIGHT};
    TILE_TO_PIPE[BOTTOM_LEFT_CORNER] = {BOTTOM, LEFT};
    TILE_TO_PIPE[BOTTOM_RIGHT_CORNER] = {BOTTOM, RIGHT};
}

PipeDirections getPipeFromTile(char tile) {
    auto it = TILE_TO_PIPE.find(tile);
    if (it != TILE_TO_PIPE.end()) {
        return it->second;
    }
    return {};
}

std::map<Coord, char> buildGrid(const std::vector<std::string>& inputLines) {
    std::map<Coord, char> grid;
    for (int y = 0; y < inputLines.size(); ++y) {
        for (int x = 0; x < inputLines[y].length(); ++x) {
            char ch = inputLines[y][x];
            if (ch != EMPTY) {
                grid[{x, y}] = ch;
            }
        }
    }
    return grid;
}

Coord findStart(const std::map<Coord, char>& grid) {
    for (const auto& pair : grid) {
        if (pair.second == START) {
            return pair.first;
        }
    }
    return UNDEFINED;
}

PipeDirections getPipeFromNeighbors(Coord coord, const std::map<Coord, char>& grid) {
    PipeDirections pipe;
    std::map<Coord, Coord> possibleNeighborOffsets = {
        {TOP, TOP},
        {RIGHT, RIGHT},
        {BOTTOM, BOTTOM},
        {LEFT, LEFT}};

    for (const auto& entry : possibleNeighborOffsets) {
        Coord dir = entry.first;
        Coord offset = entry.second;
        Coord neighborCoord = {coord.first + offset.first, coord.second + offset.second};

        auto it = grid.find(neighborCoord);
        if (it != grid.end()) {
            char neighborTile = it->second;
            PipeDirections neighborPipe = getPipeFromTile(neighborTile);

            Coord oppositeDir = {-dir.first, -dir.second};
            if (neighborPipe.count(oppositeDir)) {
                pipe.insert(dir);
            }
        }
    }
    return pipe;
}

std::vector<Coord> pathFinding(Coord start, const std::map<Coord, char>& grid) {
    std::vector<Coord> path;
    path.push_back(start);

    PipeDirections startPipe = getPipeFromNeighbors(start, grid);

    Coord previousDir = UNDEFINED;
    Coord current = UNDEFINED;

    if (!startPipe.empty()) {
        previousDir = *startPipe.begin();
        current = {start.first + previousDir.first, start.second + previousDir.second};
    } else {
        return {};
    }

    while (current != start) {
        path.push_back(current);
        char currentTile = grid.at(current);
        PipeDirections currentPipe = getPipeFromTile(currentTile);

        for (const auto& dir : currentPipe) {
            Coord oppositePrevDir = {-previousDir.first, -previousDir.second};
            if (dir != oppositePrevDir) {
                previousDir = dir;
                current = {current.first + dir.first, current.second + dir.second};
                break;
            }
        }
    }
    return path;
}

std::vector<std::string> readFile(const std::string& fileName) {
    std::vector<std::string> lines;
    std::ifstream file(fileName);
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

int solve(const std::vector<std::string>& inputLines) {
    std::map<Coord, char> grid = buildGrid(inputLines);
    Coord start = findStart(grid);
    std::vector<Coord> path = pathFinding(start, grid);
    int numPipesVisited = path.size();
    return numPipesVisited / 2;
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    initTileToPipe();

    std::vector<std::string> input = readFile("input.txt");
    std::cout << solve(input) << std::endl;

    return 0;
}
